test/test.js

```diff
@@ -131,8 +136,114 @@ describe('Service', function() {
       assert.notExists(properties.getProperty(key));
     });
   });
+
+  describe('#hasAccess()', function() {
+    it('should use the lock to prevent concurrent access', function(done) {
+      var token = {
+        granted_time: 100,
+        expires_in: 100,
+        refresh_token: 'bar'
+      };
+      var properties = new MockProperties();
+      properties.setProperty('oauth2.test', JSON.stringify(token));
+
+      mocks.UrlFetchApp.delayFunction = () => 100;
+      mocks.UrlFetchApp.resultFunction = () =>
+        JSON.stringify({
+          access_token: Math.random().toString(36)
+        });
+
+      var getAccessToken = function() {
+        var service = OAuth2.createService('test')
+            .setClientId('abc')
+            .setClientSecret('def')
+            .setTokenUrl('http://www.example.com')
+            .setPropertyStore(properties)
+            .setLock(new MockLock());
+        if (service.hasAccess()) {
+          return service.getAccessToken();
+        } else {
+          throw new Error('No access: ' + service.getLastError());
+        };
+      }.future();
+
+      Future.task(function() {
+        var first = getAccessToken();
+        var second = getAccessToken();
+        Future.wait(first, second);
+        return [first.get(), second.get()];
+      }).resolve(function(err, accessTokens) {
+        if (err) {
+          done(err);
+        }
+        assert.equal(accessTokens[0], accessTokens[1]);
+        done();
+      });
+    });
+  });
+
+  describe('#refresh()', function() {
+    /*
+      A race condition can occur when two executions attempt to refresh the
+      token at the same time. Some OAuth implementations only allow one
+      valid access token at a time, so we need to ensure that the last access
+      token granted is the one that is persisted. To replicate this, we have the
+      first exeuction wait longer for it's response to return through the
+      "network" and have the second execution get it's response back sooner.
+    */
+    it('should use the lock to prevent race conditions', function(done) {
+      var token = {
+        granted_time: 100,
+        expires_in: 100,
+        refresh_token: 'bar'
+      };
+      var properties = new MockProperties();
+      properties.setProperty('oauth2.test', JSON.stringify(token));
+
+      var count = 0;
+      mocks.UrlFetchApp.resultFunction = function() {
+        return JSON.stringify({
+          access_token: 'token' + count++
+        });
+      };
+      var delayGenerator = function*() {
+        yield 100;
+        yield 10;
+      }();
```

Comment:
If you're going to start using ES6 generators, you might want to use `let`/`const`, `() =>`, destructuring, etc.

Reply:
Maybe some day :-)